text_machina/src/metrics/token_classification.py

```diff
@@ -0,0 +1,213 @@
+    dataset = dataset.train_test_split(test_size=test_size)
+    return dataset
+
+
+def fit(
+    model: AutoModelForTokenClassification,
+    dataset: Dataset,
+    tokenizer: AutoTokenizer,
+    training_args: Dict,
+) -> None:
+    """
+    Fits a model on a dataset.
+
+    Args:
+        model (AutoModelForTokenClassification): a model.
+        dataset (Dataset): a training dataset.
+        tokenizer (AutoTokenizer): a tokenizer.
+        training_args (Dict): args to be passed to the HF's Trainer.
+    """
+    training_args = TrainingArguments(do_train=True, **training_args)
+    collator = DataCollatorForTokenClassification(tokenizer=tokenizer)
+    trainer = Trainer(
+        model,
+        train_dataset=dataset,
+        data_collator=collator,
+        args=training_args,
+    )
+    trainer.train()
+
+
+def predict(
+    model: AutoModelForTokenClassification,
+    dataset: Dataset,
+    tokenizer: AutoTokenizer,
+) -> List[List[int]]:
+    """
+    Predicts a dataset using a model.
+
+    Args:
+        model (AutoModelForTokenClassification): a model.
+        dataset (Dataset): a test dataset.
+        tokenizer (AutoTokenizer): a tokenizer.
+
+    Returns:
+        List[List[int]]: list of predicted labels for each example.
+    """
+    collator = DataCollatorForTokenClassification(tokenizer=tokenizer)
+    trainer = Trainer(
+        model,
+        data_collator=collator,
+    )
+    predictions = trainer.predict(dataset).label_ids
+    return predictions
+
+
+def eval(
+    predictions: List[List[int]],
+    references: List[List[int]],
+    label_mapping: Dict[int, str],
+) -> Dict[str, float]:
+    """
+    Evaluates using `seqeval` from HF metrics.
+
+    Args:
+        predictions (List[List[int]]): list of predictions for each example.
+        references (List[List[int]]): list of gold labels for each example.
+        label_mapping (Dict[int, str]): label mapping int to str labels.
+
+    Returns:
+        Dict[str, float] -> dictionary with metric values.
+    """
+    preds = [
+        [label_mapping[p] for (p, l) in zip(prediction, label) if l != -100]
+        for prediction, label in zip(predictions, references)
+    ]
+
+    refs = [
+        [label_mapping[l] for (_, l) in zip(prediction, label) if l != -100]
+        for prediction, label in zip(preds, references)
+    ]
+
+    preds = [pred[:-1] for pred in preds]
+
+    seqeval = evaluate.load("seqeval")
+    results = seqeval.compute(predictions=preds, references=refs)
+
+    return {
+        "precision": results["overall_precision"],
+        "recall": results["overall_recall"],
+        "f1": results["overall_f1"],
+        "accuracy": results["overall_accuracy"],
+    }
+
+
+class TokenClassificationMetric(Metric):
+    """
+    Implements a HF token classification model
+    for evaluating a mixcase dataset.
+    """
+
+    def _run(self, dataset: Dataset, **kwargs) -> pd.DataFrame:
+        if self.task_type != TaskType.MIXCASE:
```

Comment:
Would it make sense to support the boundary task type here too? we'd only need to post-process the label to work with this code

Reply:
totally! done.

Reply:
+fixed two bugs: (1) was returning label_ids from the dataset instead of the predictions, (2) missed discarding the [CLS] token.